### PR TITLE
Implement support for named ports in NetPols

### LINKED ISSUE #131: Named ports in NetPols lead to error
When using illuminatio with a NetPol containing a named port (such as [this one](https://github.com/cloudogu/k8s-security-demos/blob/e67350e8b33839e3a3516f053869401a664b2b44/2-network-policies/network-policies/3-ingress-production-allow-nosqlclient-mongo.yaml#L13)) `illuminatio build` crashes with 

```
Traceback (most recent call last):
  File "/home/myuser/.local/bin/illuminatio", line 8, in <module>
    sys.exit(cli())
  File "/usr/lib/python3/dist-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 1163, in invoke
    rv.append(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/myuser/.local/lib/python3.8/site-packages/illuminatio/illuminatio.py", line 164, in run
    ) = execute_tests(cases, orch, cri_socket)
  File "/home/myuser/.local/lib/python3.8/site-packages/illuminatio/illuminatio.py", line 212, in execute_tests
    ) = orch.ensure_cases_are_generated(core_api)
  File "/home/myuser/.local/lib/python3.8/site-packages/illuminatio/test_orchestrator.py", line 422, in ensure_cases_are_generated
    ) = self._find_or_create_cluster_resources_for_cases(cases_dict, core_api)
  File "/home/myuser/.local/lib/python3.8/site-packages/illuminatio/test_orchestrator.py", line 358, in _find_or_create_cluster_resources_for_cases
    ) = self._get_target_names_creating_them_if_missing(target_dict, api)
  File "/home/myuser/.local/lib/python3.8/site-packages/illuminatio/test_orchestrator.py", line 241, in _get_target_names_creating_them_if_missing
    rewritten_ports = self._rewrite_ports_for_host(
  File "/home/myuser/.local/lib/python3.8/site-packages/illuminatio/test_orchestrator.py", line 196, in _rewrite_ports_for_host
    port_int = int(port.replace("-", ""))
ValueError: invalid literal for int() with base 10: 'mongodb'
```

Find attached the output of `illuminatio -v DEBUG run > illuminatio.log 2>&1`
[illuminatio.log](https://github.com/inovex/illuminatio/files/5729415/illuminatio.log)

### PR DESCRIPTION
**Pull request by Phil1602**
Fixes #131 

# Changes
Illuminatio will lookup the port name during test case orchestration and will use numerical port by then. Currently the first matched pod/container's port will be used, which should be sufficient for the most cases. Note that there are some cases, which are still not yet covered and could lead to false positives for example (see 2. point in TODOs).

While testing the implementation using https://github.com/cloudogu/k8s-security-demos I have discovered some more issues.
Most of them need further investigation, but I have created some sanity checks (e.g. if =! none) to handle the errors and get at least a result. I will try to split those different related issues into actual issues to have them addressed separately. 

# Implementation
The named ports are printed out as is for the generated test cases, but are resolved later one to numerical ports. This is basically the same approach as for the pod label selectors, which are translated to actual pods or dummy pods.

# TODOs
- [x] Cleanup commits
- [x] Think about the problem that different pods with the same labelset could have ports with different number but equal names. It be nice if we would be able to actually test it, because there could be also some differences between the CNIs in their implementation/support of this "edge case".
-> I will create a followup issue therefore since this requires architectural redesign in some places.
- [x] Unit tests
- [x] e2e test cases
- [ ] Create follow-up issues for the discovered problems
- [x] Check if egress ports can also be specified using a name of a label selected pod + implement if necessary
-> Egress is currently not (fully) supported anway(?), see #65. Since namedports are rarely used in egress policies, I decided to skip this feature for now.

### CODE REVIEW

File: src/illuminatio/test_orchestrator.py [hunk 1]
```diff
@@ -207,7 +217,55 @@ def _rewrite_ports_for_host(self, port_list, services_for_host):
...
+                if p.target_port == numerical_port_without_prefix
+                or p.target_port == named_port_without_prefix
+            ]
+            if service_ports_for_port:
+                rewritten_ports[named_port] = "%s%s" % (
+                    prefix,
+                    str(service_ports_for_port[0].port),
+                )
+            else:
+                # TODO change to exception (explanation see above)
```
Comment (maxbischoff) [inline src/illuminatio/test_orchestrator.py:260]: which is the explanation?

**Review by maxbischoff — CHANGES_REQUESTED** [review]
Some minor comments and one potential issues with named ports containing a `-`. Otherwise LGTM, thanks for contributing :)

File: src/illuminatio/test_orchestrator.py [hunk 2]
```diff
@@ -181,21 +184,28 @@ def _rewrite_ports_for_host(self, port_list, services_for_host):
...
             )
+
         for port in numbered_ports:
             prefix = "-" if "-" in port else ""
             port_int = int(port.replace("-", ""))
             service_ports_for_port = [
-                p for p in service_ports if p.target_port == port_int
+                p
+                for p in service_ports
+                # TODO Services could also contain named ports. We should cross-compare here (resvoledName == numerical)
```
[the remark this answers is not in the document]
Comment (Phil1602) [reply src/illuminatio/test_orchestrator.py:207]: The comment is related to the following issue (I guess we think about different things):

We are iterating over `numbered_ports`, but it could be the case, that the NetworkPolicy itself contains a numerical port (e.g. 80), but the Service uses a name (e.g. http), which is defined within the container.
In this case we are fine, since there is a service for port 80, but the comparison (e.g. if 80 == http) would fail and let the case fail, hence we would need to resolve the service ports here first to handle each edge case correctly.

I decided to skip this and only leave a comment as this is a edge case. I guess someone will use named ports in the NetPol and Svc accordingly in general.

File: src/illuminatio/illuminatio.py [hunk 3]
```diff
@@ -264,6 +264,14 @@ def transform_results(
                 LOGGER.debug("mapped_sender_pod: %s", mapped_sender_pod)
                 LOGGER.debug("mapped_receiver_pod: %s", mapped_receiver_pod)
                 LOGGER.debug("raw_results: %s", json.dumps(raw_results))
+                if mapped_receiver_pod not in raw_results[mapped_sender_pod]:
+                    LOGGER.error(
```
[the remark this answers is not in the document]
Comment (Phil1602) [reply src/illuminatio/illuminatio.py:268]: Sounds good, yes.

I thought about refactoring this whole error handling in another gh issue (basically differentiate between success, failure(case failed) and error (prerequisites etc failed). But you are right, should at least be consistent for now.

File: src/illuminatio/test_orchestrator.py (hunk 1 above)
Comment (Phil1602) [reply src/illuminatio/test_orchestrator.py:260]: The same as for the numerical ports. Unfortunately the comment is not within the "review window":

"# TODO this was a hotfix for recipe 11, where ports 53 were allowed but not for any target,
            # resulting in test to 53 being written despite no service matching them existing.
            # That error should be handled in test generation, an exception here would be fine"

File: local_dev/start_docker.sh [hunk 4]
```diff
@@ -11,7 +11,8 @@ minikube start \
     --cni=calico \
     --container-runtime=docker \
     --host-only-cidr=172.17.17.1/24 \
-    --kubernetes-version="${KUBERNETES_VERSION}"
+    --kubernetes-version="${KUBERNETES_VERSION}" \
+    --insecure-registry=localhost:5000
```
[the remark this answers is not in the document]
Comment (Phil1602) [reply local_dev/start_docker.sh:15]: Kubelet was not able to pull the images from the internal minikube registry addon.
After adding this, it worked for me. But I did not further investigate logs besides the kubelet event, so I'm not sure whats the exact cause and why the pipeline does not have this issue.

File: src/illuminatio/test_orchestrator.py [hunk 5]
```diff
@@ -207,7 +217,55 @@ def _rewrite_ports_for_host(self, port_list, services_for_host):
...
+        for named_port in named_ports:
+            host_string_separated = str(host_string).split(":")
+            namespace = host_string_separated[0]
+            label_selector_string = (
+                host_string_separated[1] if host_string_separated[1] != "*" else ""
+            )
+
+            # Strip of the "-" and add it again after name has been resolved
+            prefix = "-" if "-" in named_port else ""
+            named_port_without_prefix = named_port.replace("-", "")
```
[the remark this answers is not in the document]
Comment (Phil1602) [reply src/illuminatio/test_orchestrator.py:237]: Good catch, thanks.

**Review by Phil1602 — COMMENTED** [review]
I will provide fixes during this week (hopefully ;) )

**Review by maxbischoff — APPROVED** [review]
